### PR TITLE
Script profile: Separate toHandlerScript for commands and states

### PR DESCRIPTION
**Pull request by florian-h05**
This allows much more fine-grained control for the script profile. 

E.g. it is now possible to mimic the behaviour of the `system:follow` profile, but apply a script transformation to the forwarded state. I discovered the need for this when trying to make a KNX channel follow an Item state, but at the same time needed to transform the Item state before sending it to KNX.

### CODE REVIEW

File: bundles/org.openhab.core.automation.module.script/src/main/resources/OH-INF/config/script-profile.xml [hunk 1]
```diff
@@ -7,14 +7,27 @@
 	<config-description uri="profile:transform:SCRIPT">
 		<parameter name="toItemScript" type="text">
 			<label>Thing To Item Transformation</label>
-			<description>The Script for transforming state updates and commands from the Thing handler to the item. The script
+			<description>The Script for transforming state updates and commands from the Thing handler to the Item. The script
```
Comment (J-N-K) [inline bundles/org.openhab.core.automation.module.script/src/main/resources/OH-INF/config/script-profile.xml:10]: Why did you change from "item" to "Item"? IMO "item" is correct and "Script/Thing" should be changed to "script/thing" in that sentence.

Comment (florian-h05) [reply bundles/org.openhab.core.automation.module.script/src/main/resources/OH-INF/config/script-profile.xml:10]: I thought that openHAB terms like Item or Thing should be written uppercase

Comment (florian-h05) [reply bundles/org.openhab.core.automation.module.script/src/main/resources/OH-INF/config/script-profile.xml:10]: How to continue here?

Comment (florian-h05) [reply bundles/org.openhab.core.automation.module.script/src/main/resources/OH-INF/config/script-profile.xml]: I have written `item` lowercase and have not changed Script/Thing yet.
I would suggest to leave it as is and probably discuss a convention across all maintainers so no one is confused anymore whether Item has to be spelled upper- or lowercase.

File: bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/profile/ScriptProfile.java [hunk 2]
```diff
@@ -92,18 +105,25 @@ public ProfileTypeUID getProfileTypeUID() {
...
+        }
+    }
+
+    private void internalFromItem(String script, Type type) {
+        String returnValue = executeScript(script, type);
+        if (returnValue != null) {
+            // try to parse the value
+            Command newCommand = TypeParser.parseCommand(handlerAcceptedCommandTypes, returnValue);
+            if (newCommand != null) {
+                callback.handleCommand(newCommand);
```
Comment (J-N-K) [inline bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/profile/ScriptProfile.java:126]: If we try to keep it as "open as possible", do we need to apply the same workaround here that we have in https://github.com/openhab/openhab-addons/pull/16263?

File: bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/profile/ScriptProfile.java [hunk 3]
```diff
@@ -92,18 +105,25 @@ public ProfileTypeUID getProfileTypeUID() {
 
     @Override
     public void onStateUpdateFromItem(State state) {
+        if (isConfigured && !toHandlerStateScript.isBlank()) {
```
Comment (J-N-K) [inline bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/profile/ScriptProfile.java:108]: Is it necessary to check `!toHandlerStateScript.isBlank()` here? `executeScript` checks `.isBlank()` anyway.

File: bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/profile/ScriptProfile.java [hunk 4]
```diff
@@ -92,18 +105,25 @@ public ProfileTypeUID getProfileTypeUID() {
...
     public void onCommandFromItem(Command command) {
-        if (isConfigured) {
-            String returnValue = executeScript(toHandlerScript, command);
-            if (returnValue != null) {
-                // try to parse the value
-                Command newCommand = TypeParser.parseCommand(handlerAcceptedCommandTypes, returnValue);
-                if (newCommand != null) {
-                    callback.handleCommand(newCommand);
-                }
+        if (isConfigured && !toHandlerCommandScript.isBlank()) {
```
Comment (J-N-K) [inline bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/profile/ScriptProfile.java:115]: See above

File: bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/profile/ScriptProfile.java [hunk 5]
```diff
@@ -43,6 +43,8 @@ public class ScriptProfile implements StateProfile {
 
     public static final String CONFIG_TO_ITEM_SCRIPT = "toItemScript";
     public static final String CONFIG_TO_HANDLER_SCRIPT = "toHandlerScript";
+    public static final String CONFIG_TO_HANDLER_COMMAND_SCRIPT = "toHandlerCommandScript";
```
Comment (J-N-K) [inline bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/profile/ScriptProfile.java:46]: If we change the naming anyway: Wouldn't `stateFromItemScript` and `commandFromItemScript` be more clear? I'm not sure here, it just came to my mind.

**Review by J-N-K — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/profile/ScriptProfile.java (hunk 5 above)
Comment (florian-h05) [reply bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/profile/ScriptProfile.java:46]: Agreed! Having `toItem` and `fromItem` IMO is better to understand for the user as `toHandler` and `toItem`, because I am not sure whether most users actually know what the/a (Thing) handler is.

File: bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/profile/ScriptProfile.java (hunk 3 above)
Comment (florian-h05) [reply bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/profile/ScriptProfile.java:108]: I haven't noticed that, it is not neccassary to double check.

File: bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/profile/ScriptProfile.java (hunk 2 above)
Comment (florian-h05) [reply bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/profile/ScriptProfile.java:126]: Probably yes, but I am not sure whether we want to "distribute" that work-around across the code or if it would be better to wait for a solution of the underlying issue/problem.

Comment (florian-h05) [reply bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/profile/ScriptProfile.java:126]: While having a look at `ProfileCallbackImpl`, the problem seems to be that `CommunicationManager::toAcceptedCommand` does not accept `OpenClosedType`, which IMO is correct as you normally do not command contact Items.

The question is, how to solve that.
I would think the cleanest solution is to add a method like `CommunicationManager::toAcceptedState` and to introduce a method called `handleState` or `handleUpdate` to the `ProfileCallback`. 
Then all profiles could call `handleUpdate` instead of `handleCommand` for state updates from the Item to the Thing.

**Comment by florian-h05** [comment]
@J-N-K I don't want to be pushy, but it would be great to get this PR merged soon.

File: bundles/org.openhab.core.automation.module.script/src/main/resources/OH-INF/config/script-profile.xml [hunk 6]
```diff
@@ -11,12 +11,25 @@
...
 				discard the commands and not pass them through.</description>
 			<limitToOptions>false</limitToOptions>
 		</parameter>
+		<parameter name="stateFromItemScript" type="text">
+			<label>Item State To Thing Transformation</label>
+			<description>The Script for transforming states from the item to the Thing handler. The script may return null to
+				discard the states and not pass them through.</description>
+			<limitToOptions>false</limitToOptions>
+		</parameter>
+		<parameter name="toHandlerScript" type="text">
```
Comment (J-N-K) [inline bundles/org.openhab.core.automation.module.script/src/main/resources/OH-INF/config/script-profile.xml:26]: Maybe show this one as "advanced"? That hides it for new users, while keeping it for this that already have it configured.

Side note: It would be great if Main UI would automatically show "advanced" parameters if something is configured there. This would also be helpful e.g. for HTTP binding things.

**Review by J-N-K — COMMENTED** [review]
Otherwise LGTM

**Review by J-N-K — APPROVED** [review]
Thanks!

File: bundles/org.openhab.core.automation.module.script/src/main/resources/OH-INF/config/script-profile.xml (hunk 6 above)
Comment (florian-h05) [reply bundles/org.openhab.core.automation.module.script/src/main/resources/OH-INF/config/script-profile.xml:26]: > Side note: It would be great if Main UI would automatically show "advanced" parameters if something is configured there. This would also be helpful e.g. for HTTP binding things.

There we go: https://github.com/openhab/openhab-webui/pull/2313

**Comment by florian-h05** [comment]
@openhab/jruby-maintainers FYI in case you need to adjust documentation.

**Comment by florian-h05** [comment]
Yeah, good idea.
